src/ObjCRuntime/Runtime.cs

```diff
@@ -1437,38 +1437,36 @@ static IntPtr CreateNSObject (IntPtr type_gchandle, IntPtr handle, NSObject.Flag
 				return null;
 
 			var obj = TryGetNSObject (ptr, evenInFinalizerQueue: false);
-			T? o;
-
-			if (obj is null) {
-				// Try to get the managed type that correspond to this exact native type
-				IntPtr p = Class.GetClassForObject (ptr);
-				// If unknown then we'll get the Class that Lookup to NSObject even if this is not NSObject.
-				// We can use this condition to fallback on the provided (generic argument) type
-				Type target_type;
-				if (p != NSObjectClass) {
-					target_type = Class.Lookup (p);
-					if (target_type == typeof(NSObject))
-						target_type = typeof(T);
-					else if (typeof (T).IsGenericType)
-						target_type = typeof(T);
-					else if (target_type.IsSubclassOf (typeof(T))) {
-						// do nothing, this is fine.
-					} else if (Messaging.bool_objc_msgSend_IntPtr (ptr, Selector.GetHandle ("isKindOfClass:"), Class.GetHandle (typeof (T)))) {
-						// If the instance itself claims it's an instance of the provided (generic argument) type,
-						// then we believe the instance. See bug #20692 for a test case.
-						target_type = typeof(T);
-					}
-				} else {
-					target_type = typeof(NSObject);
+
+			// First check if we got an object of the expected type
+			if (obj is T o)
+				return o;
```

Comment:
Will this cause issue if for example we have an initial type X and then we try get instance of type Y, but Y : X due to inheritance.

I am thinking `NSDictionary<K, V> : NSDictionary`

Reply:
No, if the first instance is `NSDictionary`, and then you want an `NSDictionary<K, V>`, then this check will fail and we'll create a new managed `NSDictionary<K, V>` instance (previously we'd throw an InvalidCastException for this scenario).

For the reverse case, the first instance is `NSDictionary<K, V>` and you want an `NSDictionary`, you'll get the `NSDictionary<K, V>` instance (it was this way before this PR, and will be after too).